File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/EnvoyStreamEmitter.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream
import java.nio.ByteBuffer

class EnvoyStreamEmitter(
  private val stream: EnvoyHTTPStream
) : StreamEmitter {

  override fun sendData(byteBuffer: ByteBuffer): StreamEmitter {
    stream.sendData(byteBuffer, false)
    return this
  }

  override fun close(trailers: Map<String, List<String>>) {
    stream.sendTrailers(trailers)
  }

  override fun close(byteBuffer: ByteBuffer) {
    stream.sendData(byteBuffer, true)
  }

  override fun cancel() {
    stream.cancel()
  }
}

File: mobile/library/swift/src/EnvoyStreamEmitter.swift
import Foundation

/// Default implementation of the `StreamEmitter` interface.
@objcMembers
final class EnvoyStreamEmitter {
  private let stream: EnvoyHTTPStream

  init(stream: EnvoyHTTPStream) {
    self.stream = stream
  }
}

extension EnvoyStreamEmitter: StreamEmitter {
  func sendData(_ data: Data) -> StreamEmitter {
    self.stream.send(data, close: false)
    return self
  }

  func sendMetadata(_ metadata: [String: [String]]) -> StreamEmitter {
    self.stream.sendMetadata(metadata)
    return self
  }

  func close(trailers: [String: [String]]) {
    self.stream.sendTrailers(trailers)
  }

  func close(data: Data) {
    self.stream.send(data, close: true)
  }

  func cancel() {
    _ = self.stream.cancel()
  }
}
